### PR TITLE
[GSoC] Implement the CSV importer

### LINKED ISSUE #6772: Feature: Import CSV/TSV files
Anki Desktop allows the import of CSV/TSV, and makes use of the auto-detection of CSV files which is built into Python.

Splintering off this task from #3517 as this is close to complete locally.

There will likely be follow up support tickets which would best be associated with a specific ticket rather than an 'Epic'. This is a significant port of Python and C code, and there is a large likelihood for bugs.

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Purpose / Description
On top of #12072

## Fixes
Closes #6772
Closes #9437

## Approach

This implements the base for using upstream's HTML/Js pages directly on AnkiDroid.

I've used https://github.com/krmanik/Anki-Android/pull/23 as base of this PR. I wouldn't have managed to do it without all the previous work of Mani and Damien, so I thank them very much.

As discussed on https://github.com/krmanik/Anki-Android/pull/23, to use an Anki HTML page, we should be able to intercept POST requests. After some proof of concepts made by Mani, the conclusion was that hosting a local server, like the desktop application, may be most suitable approach.

To create the local server, I've used the same approach of https://github.com/krmanik/Anki-Android/pull/23 and added NanoHTTPD, a java server library, as dependency. I don't have experience with any Java server libraries, so I can't tell with confidence if this is *the best library*, but it seemed solid to me and already had a working model on Mani's PR, so using it looked like the best choice for the moment.

Besides the server, this PR implements the base classes for creating AnkiPages. Each page can easily extend `PageFragment` and define what it needs. I've tried using the new Statistics, Card Info, Deck Options pages and they work well. Try https://github.com/BrayanDSO/Anki-Android/tree/csvImporterOtherPages if you want to test them (the deck options screen may need some extra work).

Finally, the CSV importer page and its endpoints are implemented and can be used with the new backend.

**PENDING (for future PRs):**
- Close the Import dialog after the taps "OK" on the progress dialog
- Inject CSS to match the themes' colors
- ~~Maybe hide the file path with CSS/Js, because it may be big and there isn't much horizontal space in phone screens on portrait mode~~ (already on upstream)
- Not for my GSoC project, but the other screens could be implemented too, as they are mostly done, so other people could test/work on them as well

## How Has This Been Tested?

NOTICE: this needs an updated WebView (77+) to work correctly (see https://github.com/ankidroid/Anki-Android/issues/6772#issuecomment-1209212058 and the comments below for discussion), which shouldn't be a problem as the project's minSdk (21) supports it. But if you are testing on a emulator, be aware that the WebView shipped on some older SDK versions may be an old version, so either update the WebView or use a newer SDK.

NOTICE 2: there may be a flickering issue after remapping fields, but that was already solved on upstream (https://github.com/ankitects/anki/pull/2005)

1. Enable the new backend (Settings > Advanced)
2. Deck picker > Three dot menu > Import > Text file
3. Select a CSV file (I've tested with Anking, pretty heavy and popular deck)
    - Here's a mini version (23KB) if anyone wants to test it: [MiniAnking.txt](https://github.com/ankidroid/Anki-Android/files/9383917/MiniAnking.txt)
    - And here's a slightly cut version (GitHub only allows files up to 25MB): 
[Anking.txt](https://github.com/ankidroid/Anki-Android/files/9383942/Anking.txt)
4. Configure anything on the CSV importer
5. Tap Import

https://user-images.githubusercontent.com/69634269/185787227-5df94718-f6d4-472a-8301-159f4628110d.mp4

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [X] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt [hunk 1]
```diff
@@ -46,6 +50,22 @@ fun DeckPicker.importApkgs(apkgPaths: List<String>) {
     }
 }
 
+fun PagesActivity.importCsvRaw(input: ByteArray): ByteArray? {
+    launchCatchingTask {
+        val output = withProgress(
+            extractProgress = { if (progress.hasImporting()) { text = progress.importing } },
+            op = { withCol { (this as CollectionV16).importCsvRaw(input) } }
```
Comment (dae) [inline AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:57]: This may come down to personal preference, but in the other cases I've used withProgress, it's with a trailing closure, eg:

```
    val output = deckPicker.withProgress(
        extractProgress = {
            if (progress.hasNormalSync()) {
                text = progress.normalSync.run { "$added\n$removed" }
            }
        },
        onCancel = ::cancelSync
    ) {
        withCol { newBackend.syncCollection(auth) }
    }
```

I think CSV importing also supports cancellation, which you can achieve with the following line:

```
        onCancel = { backend -> backend.setWantsAbort() }
```

The apkg importer also supports cancellation, but I forgot to add it.

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt [hunk 2]
```diff
@@ -46,6 +50,22 @@ fun DeckPicker.importApkgs(apkgPaths: List<String>) {
...
+        )
+        MaterialDialog(this@importCsvRaw).show {
+            message(text = summarizeReport(col.tr, ImportResponse.parseFrom(output)))
+            positiveButton(R.string.dialog_ok) {
+                this@importCsvRaw.finish()
+            }
+        }
+    }
+    return null
```
Comment (dae) [inline AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:66]: This function is declared as returning a ByteArray?, but it currently always returns null, as the computing happens in a suspend function. In this particular case the return value does not matter, but it may change in the future, so it would be cleaner to return the result of the importCsvRaw() call. Are the handlers in AnkiServer called on a background thread? If so, you could replace the launchCatchingTask call with a runBlocking call, though you'd need to replicate the error handling that's done in launchCatchingTask (maybe we need a runCatchingTask?)

**Review by dae — COMMENTED** [review]
Nice!

**Review by krmanik — COMMENTED** [review]
Should we move last two commits to separate PR for better diff in this PR?

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 1 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:58]: Added the cancel op. Is CSV import undoable as well?

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 2 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:67]: The server runs each request on a different thread. I've tried using `runBlocking`, but this shows up

`java.lang.RuntimeException: Can't create handler inside thread Thread[NanoHttpd Request Processor (#4),5,main] that has not called Looper.prepare()`

Which means that I should either use a HandlerThread or something I don't know on the Coroutines library. I'll do some research and try some things.

**Comment by BrayanDSO** [comment]
`Should we move last two commits to separate PR for better diff in this PR?`

Good idea

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 2 above)
Comment (krmanik) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:67]: If the server run code move to AnkiActivity then it will run only one time, so for the session of AnkiDroid run same server instance will be reused.

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:67]: As the pages are Fragments, you can keep the same PagesActivity, and therefore the same server instance, by just replacing the current fragment.

If we should make the server a singleton and keep it alive as long as there is an activity running, I don't think that now is the best time to decide it. This PR only adds one screen, so for the current scope, there is no reason for it.

When we do add the other screens, it still will need some thought/research/profiling if the performance cost of creating a new server instance is worth the cost of having enabled all the time. Especially because those screens (Statistics, Deck options, Card info) may not be used frequently, or used at all, by the user.

Comment (krmanik) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:67]: I agree with running the server when required instead of running of the beginning of running AnkiDroid.

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:67]: Okay, I've got some progress. The problem is the dialogs. As they are running on a thread that doesn't have a looper to send their Draw messages, and runBlocking uses the same thread, they end up throwing. 

For the report dialog, any of these approaches below work.

```kotlin
val h = Handler(Looper.getMainLooper())
    val r = java.lang.Runnable {
        MaterialDialog(this@importCsvRaw).show {
            message(text = summarizeReport(col.tr, ImportResponse.parseFrom(output)))
            positiveButton(R.string.dialog_ok) {
                this@importCsvRaw.finish()
            }
        }
    }
    h.post(r)
```
```kotlin
runOnUiThread {
        MaterialDialog(this@importCsvRaw).show {
            message(text = summarizeReport(col.tr, ImportResponse.parseFrom(output)))
            positiveButton(R.string.dialog_ok) {
                this@importCsvRaw.finish()
            }
        }
    }
```

For the Progress dialog, I believe that I should make a `withProgressBlocking` method that runs on the Ui Thread

**Comment by mikehardy** [comment]
Maintainer note (or, other maintainer note, since dae and krmanik are both here already): I was *really* curious about both the localhost server *and* the HTML/JS from the backend work and have also participated in PR 23 on krmanik's fork while this was developed. 

TL;DR: localhost server really seems like the best approach after testing, and I'm super-excited to get HTML/JS support from the backend. This PR is pretty amazing in the forward progress it represents for AnkiDroid

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 2 above)
Comment (dae) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:67]: withContext() would be more natural with coroutines I think. I just noticed that the code was passing a col handle into AnkiServer, which means access to the collection is unguarded for most of the requests. How about this instead:

```diff
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt b/AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt
index 96d3301f0..751d5e6f1 100644
--- a/AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt
@@ -29,6 +29,9 @@ import com.ichi2.libanki.exportCollectionPackage
 import com.ichi2.libanki.importAnkiPackage
 import com.ichi2.libanki.importer.importCsvRaw
 import com.ichi2.libanki.undoableOp
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withContext
 import net.ankiweb.rsdroid.Translations
 
 fun DeckPicker.importApkgs(apkgPaths: List<String>) {
@@ -50,10 +53,15 @@ fun DeckPicker.importApkgs(apkgPaths: List<String>) {
     }
 }
 
-fun PagesActivity.importCsvRaw(input: ByteArray): ByteArray? {
-    launchCatchingTask {
+
+suspend fun PagesActivity.importCsvRaw(input: ByteArray): ByteArray {
+    return withContext(Dispatchers.Main) {
         val output = withProgress(
-            extractProgress = { if (progress.hasImporting()) { text = progress.importing } },
+            extractProgress = {
+                if (progress.hasImporting()) {
+                    text = progress.importing
+                }
+            },
             onCancel = { backend -> backend.setWantsAbort() },
             op = { withCol { (this as CollectionV16).importCsvRaw(input) } }
         )
@@ -63,8 +71,8 @@ fun PagesActivity.importCsvRaw(input: ByteArray): ByteArray? {
                 this@importCsvRaw.finish()
             }
         }
+        output
     }
-    return null
 }
 
 private fun summarizeReport(tr: Translations, output: ImportResponse): String {
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/pages/AnkiServer.kt b/AnkiDroid/src/main/java/com/ichi2/anki/pages/AnkiServer.kt
index a228f7c37..dd7b6160b 100644
--- a/AnkiDroid/src/main/java/com/ichi2/anki/pages/AnkiServer.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/pages/AnkiServer.kt
@@ -17,18 +17,19 @@
 
 package com.ichi2.anki.pages
 
+import com.ichi2.anki.CollectionManager.withCol
 import com.ichi2.anki.importCsvRaw
 import com.ichi2.libanki.*
 import com.ichi2.libanki.importer.getCsvMetadataRaw
 import com.ichi2.libanki.stats.*
 import fi.iki.elonen.NanoHTTPD
+import kotlinx.coroutines.runBlocking
 import timber.log.Timber
 import java.io.ByteArrayInputStream
 
 class AnkiServer(
     hostname: String?,
     port: Int,
-    val col: CollectionV16,
     val activity: PagesActivity
 ) : NanoHTTPD(hostname, port) {
 
@@ -43,20 +44,10 @@ class AnkiServer(
 
         if (session.method == Method.POST) {
             Timber.d("POST: Requested %s", uri)
-            val bytes = getSessionBytes(session)
+            val inputBytes = getSessionBytes(session)
             if (uri.startsWith(ANKI_PREFIX)) {
-                val data: ByteArray? = when (uri.substring(ANKI_PREFIX.length)) {
-                    "i18nResources" -> col.i18nResourcesRaw(bytes)
-                    "getGraphPreferences" -> col.getGraphPreferencesRaw()
-                    "setGraphPreferences" -> col.setGraphPreferencesRaw(bytes)
-                    "graphs" -> col.graphsRaw(bytes)
-                    "getNotetypeNames" -> col.getNotetypeNamesRaw(bytes)
-                    "getDeckNames" -> col.getDeckNamesRaw(bytes)
-                    "getCsvMetadata" -> col.getCsvMetadataRaw(bytes)
-                    "importCsv" -> activity.importCsvRaw(bytes)
-                    "getFieldNames" -> col.getFieldNamesRaw(bytes)
-                    "cardStats" -> col.cardStatsRaw(bytes)
-                    else -> { Timber.w("Unhandled Anki request: %s", uri); null }
+                val data: ByteArray? = runBlocking {
+                    handlePostRequest(uri.substring(ANKI_PREFIX.length), inputBytes)
                 }
                 return newChunkedResponse(data)
             }
@@ -64,6 +55,22 @@ class AnkiServer(
         return newFixedLengthResponse(null)
     }
 
+    private suspend fun handlePostRequest(methodName: String, bytes: ByteArray): ByteArray? {
+        return when (methodName) {
+            "i18nResources" -> withCol { newBackend.i18nResourcesRaw(bytes) }
+            "getGraphPreferences" -> withCol { newBackend.getGraphPreferencesRaw() }
+            "setGraphPreferences" -> withCol { newBackend.setGraphPreferencesRaw(bytes) }
+            "graphs" -> withCol { newBackend.graphsRaw(bytes) }
+            "getNotetypeNames" -> withCol { newBackend.getNotetypeNamesRaw(bytes) }
+            "getDeckNames" -> withCol { newBackend.getDeckNamesRaw(bytes) }
+            "getCsvMetadata" -> withCol { newBackend.getCsvMetadataRaw(bytes) }
+            "importCsv" -> activity.importCsvRaw(bytes)
+            "getFieldNames" -> withCol { newBackend.getFieldNamesRaw(bytes) }
+            "cardStats" -> withCol { newBackend.cardStatsRaw(bytes) }
+            else -> { Timber.w("Unhandled Anki request: %s", methodName); null }
+        }
+    }
+
     private fun getSessionBytes(session: IHTTPSession): ByteArray {
         val contentLength = session.headers["content-length"]!!.toInt()
         val bytes = ByteArray(contentLength)
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt b/AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt
index 3a9313428..6bc7b70c9 100644
--- a/AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt
@@ -50,7 +50,7 @@ class PagesActivity : AnkiActivity() {
         WebView.setWebContentsDebuggingEnabled(BuildConfig.DEBUG)
 
         // Load server
-        ankiServer = AnkiServer(HOST_NAME, port, col as CollectionV16, this)
+        ankiServer = AnkiServer(HOST_NAME, port, this)
         Timber.i("Starting server on $HOST_NAME:$port")
         ankiServer.start()
 
```

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 1 above)
Comment (dae) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:58]: Ah, in that case it shouldn't be needed here either. Sorry :-) And yes, I think csv import is cancellable.

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 2 above)
Comment (dae) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:67]: (exceptions still need to be handled inside the runBlocking call or via a new helper. Most of the routines will be running on a background thread, so when reporting errors you'd probably want another withContext(main))

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 1 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt:62]: Made it undoable. Thanks a lot!

File: AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt (hunk 2 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt]: Lovely, lovely, lovely. Works like a charm now. Thank you so much!

**Comment by BrayanDSO** [comment]
Updated the PR's description with a new video, now showing cancelling and undoing the import as well.

**Review by david-allison — APPROVED** [review]
The future is now! Incredible work

**Comment by BrayanDSO** [comment]
Only squashed/simplified the last commits with the last push force

File: AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt [hunk 3]
```diff
@@ -49,26 +50,64 @@ import kotlin.coroutines.suspendCoroutine
...
+        showError(this, exc.localizedMessage!!)
+    } catch (exc: Exception) {
+        Timber.e("caught: %s %s", exc, extraInfo)
+        showError(this, exc.toString())
+    }
+    return null
+}
+
+/**
+ * Calls [runBlocking] while catching errors with [runCatchingTask]
```
Comment (dae) [inline AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:76]:
```suggestion
 * Calls [runBlocking] while catching errors with [runCatchingTask].
 * This routine has a niche use case - it allows us to integrate coroutines into nanohttpd, which runs
 * request handlers in a synchronous context on a background thread. In most cases, you will want
 * to use [FragmentActivity.launchCatchingTask] instead.
```

File: AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt [hunk 4]
```diff
@@ -49,26 +50,64 @@ import kotlin.coroutines.suspendCoroutine
...
+): T? {
+    return runBlocking {
+        runCatchingTask(errorMessage) { block() }
+    }
+}
+
+/**
+ * Calls [withContext] while catching errors with [runCatchingTask]
+ */
+suspend fun <T> FragmentActivity.withContextCatching(
```
Comment (dae) [inline AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:90]: This appears to be unused?

**Review by dae — COMMENTED** [review]
Nice work on adding undo!

**Comment by dae** [comment]
> Maybe hide the file path with CSS/Js, because it may be big and there isn't much horizontal space in phone screens on portrait mode

Already coming: https://github.com/ankitects/anki/commit/2e2fc799e26695c4d487af60a185c6cb4e20017e

File: AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt (hunk 4 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt]: It ended not being necessary and I forgot to remove it. Done.

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/AnkiServer.kt [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            "i18nResources" -> withCol { newBackend.i18nResourcesRaw(bytes) }
+            "getGraphPreferences" -> withCol { newBackend.getGraphPreferencesRaw() }
+            "setGraphPreferences" -> withCol { newBackend.setGraphPreferencesRaw(bytes) }
+            "graphs" -> withCol { newBackend.graphsRaw(bytes) }
+            "getNotetypeNames" -> withCol { newBackend.getNotetypeNamesRaw(bytes) }
+            "getDeckNames" -> withCol { newBackend.getDeckNamesRaw(bytes) }
+            "getCsvMetadata" -> withCol { newBackend.getCsvMetadataRaw(bytes) }
+            "importCsv" -> withCol { newBackend.importCsvRaw(bytes) }
+            "getFieldNames" -> withCol { newBackend.getFieldNamesRaw(bytes) }
+            "cardStats" -> withCol { newBackend.cardStatsRaw(bytes) }
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/AnkiServer.kt:69]: Looks like there is a lot of fun in store as soon as this is merged

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+         * as arguments of the [PageFragment] that will be opened
+         */
+        const val EXTRA_PAGE_ARGS = "pageArgs"
+        /**
+         * Extra key of [PagesActivity]'s intent that must be included and
+         * hold the name of an [Anki HTML page](https://github.com/ankitects/anki/tree/main/ts)
+         */
+        const val EXTRA_PAGE_NAME = "pageName"
+
+        const val HOST_NAME = "127.0.0.1"
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt:101]: this is more for learning, but would localhost be safer? Imagining an IPv6 device where `::1` is the right thing but localhost should still map?

**Review by mikehardy — APPROVED** [review]
I don't have anything actually blocking this. I'm most happy about the monumental collaboration here on all levels

David spearheading then everyone else working to get Kotlin going
Damien and Mani's efforts to get the new version of the backend up and plumbing Anki JS through
Brayan, all the work here

**Comment by BrayanDSO** [comment]
> My only question now is that I think this may orphan some of the previous stuff, specifically this? https://github.com/ankidroid/Anki-Android/blob/main/AnkiDroid/src/main/java/com/ichi2/anki/importer/CsvMapping.kt

I've dropped the commits that removed all of the legacy CSV implementation in order to keep a better diff

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt (hunk 6 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt:101]: Let me try this while my PC is still on

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt]: Okay, works here

Comment (dae) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt]: FWIW, the desktop used to use 'localhost', but switched over to '127.0.0.1', as some machines appeared to have broken host files where localhost wasn't defined, causing the app to fail to load. Might be rare/not be a problem on Android though.

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/PagesActivity.kt]: I could easily see that happening on desktop but on android that would require rooting the phone, which is getting pretty difficult these days. I am willing to take the gamble a person willing to run a rooted phone at this point will have their hosts file in order (in fact, I'd wager most of the time it is specifically so they can run an adblocker that uses the hosts file to blackhole things and they'll be really hosts-file-aware.

**Review by krmanik — APPROVED** [review]
I have pulled the PR locally and tested it on an actual device. I have tested import feature using the following files.
1. Anking.txt, provided by the Brayan
    File Size: 24 MB
    Lines in file: 20k
    Note Type: Basic 
2. Tatoeba Chinese English sentences pair
     File Size: 5 MB
     Lines in file: 63k
     Note Type: Basic
     I have interrupted import after 20k notes to test interrupt. 

I am happy with the result of the import-csv feature. But the only issue I am seeing is that only the text file is selectable to import, other file (csv and tsv) is not selectable. I renamed files to test import. 
So, may be follow up PR for the making csv and tsv selectable should be fine.

</img>

This is one of the best features coming to AnkiDroid. Thanks to @BrayanDSO.

**Comment by mikehardy** [comment]
Let's merge this to unblock all the stuff piled up on it (multiple follow-ons, that is), and the csv file selection issue can be a follow-on as well

**Comment by BrayanDSO** [comment]
> I am happy with the result of the import-csv feature. But the only issue I am seeing is that only the text file is selectable to import, other file (csv and tsv) is not selectable.

lol. Learned something today. I'd thought that Android imports were based on MIME types, so if it works with `.txt`, renaming it to `.csv` should work, which means I'm probably missing something.

But that is easy to solve anyway
